### ISSUE TITLE
Wiki clean up

### ISSUE THREAD

**Issue by czyzby** [czyzby/gdx-lml#10]
- [x] Update all examples.
- [x] Remove syntax page.
- [x] Adds tips and tricks.
- [x] Step-by-step tutorial: a simple `gdx-lml-vis` application.
- [x] Separate old tutorial into multiple parts.

**Comment by czyzby** [comment]
Removed obsolete syntax page with pre-`1` LML version tags. Syntax page is redundant with `DTD` around and requires a lot of work to finish and keep up to date.

All examples are updated to use the new `1.6` macro markers. Tips & tricks page was added in addition to the previous tutorial.

**Comment by jpsgamboa** [comment]
On the tutorial, Chapter 6. Views: actor injection.
If i'm correct, this is the first time we're showed how to actually "display" our templates and the wording is quite heavy (imo).  A title like Displaying your view (actor injection), might be friendlier.

The first paragraph of this chapter is confusing for someone new to Lml or libGDX in general (it took me a few reads to actually interiorize what I was reading).

I'd suggest making it clearer the following way:

**Introduction**
First, I'd suggest repeating here how LML fits into a libGDX application for context - You create your templates, `LmlParser` will parse them and create `Actors`, which you can then add to your `Stage`.

For this, there are **3**(is this correct?) options we can use LML templates. 

  1. We can extend `AbstractLmlView`, parse and instantiate it use calling `LmlParser#createView`. An `AbstractLmlView` expects a stage to be passed in and takes "control" of it, in the sense that it calls `Stage#act()`, `Stage#draw()`, `Stage#resize()`, `Stage#dispose()`.. for us. 
     I feel like this should be presented first, as it provides the most functionality out of the box (with LmlAction and LmlActor annotations, etc);
     `code snippet`

  2. Alternatively, we can use LmlParser#fillStage(FileHandle, Stage)... which only adds the actors to    your stage;
     `code snippet`

  3. Finally, we can use LmlParser to create Actors, and do whatever with them via LmlParser#parseTemplate(FileHandle).
     `code snippet`

**LmlView**
From here, I believe `#createView` deserves a dedicated _sub-chapter_ to show off all it's functionality.

After writing this, I realize what this chapter needs are sub-chapters. There is too much important information condensed here, and it's hard to follow and find.

Bonus question:
I have a LmlView created with `#createView`. 
If I want to display a dialog that would get called by my Java code (for example to display a warning).
For this I would have a custom `<:macro alias="warning dialog">` or a simple template with the dialog (need your suggestion here).

And then, when my java code wants to display the dialog, it would call LmlParser#fillStage("dialogtemplate.lml", LmlView$getStage())?

In this case, it's better to have a dedicated template than a custo macro right?

**Comment by czyzby** [comment]
You're absolutely right about chapters, I'll divide the tutorial into separate pages and try to address the issues to mentioned.

About the macro - first of all, tag names cannot have whitespaces, so I suggest using `<:macro alias="warningDialog">`. A separate dialog template is the way to go - and your loading method seems to be OK. Macros are parsed *immediately* during template parsing, so you cannot do something like that (yet!):

```<button>
  <onClick>
    <dialog> ... </dialog>
  </onClick>
</button>
```
I actually thought about something like this a while back. There were some problems with the implementation (not sure what stopped me other than lack of time or laziness right now), but if you think that's worth, I can add some dedicated tags.

**Comment by czyzby** [comment]
By the way, @jpsgamboa, have you tried going though `gdx-lml-tests` or `gdx-lml-vis-tests` code? Do you consider these projects useful and pretty clean?

**Comment by jpsgamboa** [comment]
Thanks for the heads up about the whitespace.
I'm mainly concerned about repetition. If the user already has a custom macro for a certain dialog, but this time he wants it to be called from the java code, he would have to duplicate the dialog in a template.
----
Certainly, I run `gdx-lml-vis-tests` every time to seek inspiration or learn syntax. So in terms of usefulness, they are brilliant and I don't think there would be a better way to showcase LML. It's definitely worth definitely worth maintaining them.

The sources themselves could be updated to use DTD and other new features because people look at the source to learn, and consistency is key.

**Comment by jpsgamboa** [comment]
This gives me the ideia that `gdx-lml-tests` and `gdx-lml-vis-tests ` could be expanded as a general introduction to LML in the following ways:

- When the app in launched, a few screens can be showed providing some context of how the library can be used, what are the benefits, etc. This should ofcourse be created using LML, and the user would be advised to look at the code to see how it works.

- Secondly, the user ends up in the examples section (what exists today).

The first half would work more like any other application (maybe a splash screen, a main menu, so on) and the source code for this could be used other users trying to adopt LML.

The second half would serve the purpose it does today, showcase all the possibilities in a compact manner, and provide a way for the user to try and edit some LML templates.

**Comment by czyzby** [comment]
I already updated many templates to show the new named attributes in macros - it should be enough to get an idea how to create valid XML files. Don't forget that DTD is optional and one of the main reasons to use LML over plain Java is *less* boilerplate. Sure, content assist is great, but in my personal projects (especially if they are small) I'd still rather go with `<:if {value} != 0>` than more verbose `<:if test="{value} != 0">`. The example project is also supposed to show that "hey, if you don't like XML, you can throw some code without quotations and weird syntax and it will still work". Remember that LML matches LibGDX API, so anyone fluent enough with `Scene2D` should have no problems looking up attributes' names - they are simply the names of Java methods with `set` stripped.

Lack of DTD usage examples is a good point though. I think about adding a separate project with some basic DTD support can be a good idea. Actually, a wiki article with some screenshots that walks you through creating a simple application using `gdx-lml-vis` (with explanation how to use add it to your project, how to hook up DTD, etc.) could be useful.

Adding more screens to `tests` projects is a tempting idea, but also requires some work and will probably make the projects considerably more complex. There are already some tutorials here and there, flooding the user with text isn't really the goal of `tests` - they are supposed to show you the tags usage. They are something you can quickly run and test your ideas. I'm not sure if we should change that.

By the way, `LmlView` and other screen maintenance utilities weren't the goal of LML in the first place - this is a fairly new feature. Do you think that LML should have some basic view manager that you can use? Something to pair `LmlViews` with, like what LibGDX `Game` is to `Screen`.

**Comment by czyzby** [comment]
> I'm mainly concerned about repetition. If the user already has a custom macro for a certain dialog, but this time he wants it to be called from the java code, he would have to duplicate the dialog in a template.

That's true, but you wouldn't have to duplicate code with macros either if the listeners attachment was already there. Consider this macro:
```
<:macro alias="showError">
  <onClick><dialog>Don't click that!</dialog></onClick>
</:macro>
```
This would be defined *once* and you could just use it everywhere like that: `<button><:showError/></button>`. Actually, using an action that parses a template each time the listener is invoked might be more unpleasant, as you would have to keep it as a global action (to make it accessible from all templates) and create some sort of caching mechanism by hand to prevent the dialog from being displayed multiple times.

Anyway, if you have any more questions about this topic, please use #13 issue.

**Comment by jpsgamboa** [comment]
I understand your points.

I don't think LML _needs_ it, if the current usage mechanism is clear. I had some difficulties understanding how LML (and LMLView in particular) would fit in my app initially, but it became clear after some tutorial reads (and your answers). I believe the wiki modifications we've discussed would cover this.

Sometimes having too many options (like using LML to manage everything vs using to create a view and handle the stage vs using it only to parse some Actor) can become too confusing.

That being said, at first sight it doesn't seem (I might be very wrong) that creating such a manager would be complicated, and LML would go full circle, covering almost all aspects of UI development.

**Comment by czyzby** [comment]
It would be very simple.

**Comment by czyzby** [comment]
@jpsgamboa, could you go through [a simple tutorial](https://github.com/czyzby/gdx-lml/wiki/Your-first-LML-application)? It might not be advanced enough for you by now, but I'd appreciate your input.

**Comment by czyzby** [comment]
@jpsgamboa, I separated the old tutorial into sections. [Parsing section](https://github.com/czyzby/gdx-lml/wiki/Parsing) was added as you suggested - it explains how to actually *parse* LML templates and should be treated as the introduction to [actor injection](https://github.com/czyzby/gdx-lml/wiki/Actor-injection). Not much content has changed though, I only updated it with the new syntax informations, etc.

**Comment by jpsgamboa** [comment]
Looks great! 
Few things that raised an eyebrow:
- The example features VisUI so when you create the LmlParser, would be nice to have a link to a non VisUI example. (it's not much different, but would allow someone who doesn't care about VisUI to follow the tutorial without getting lost)
- In Global actions, you advise the reader to create a `actions` package to place the `ActionContainer`, but it reads like it is mandatory to create such a package, which might confuse some people. I think saying simply

> Create a new class (e.g. Global)..

would be enough.
- Later in this chapter you use the `{element}` but make no reference to what it means.

Overall nice guide!

**Comment by czyzby** [comment]
Added `gdx-lml` equivalent of `LmlParser` creation ([above this section](https://github.com/czyzby/gdx-lml/wiki/Your-first-LML-application#your-first-view)), more details about the `ActionContainer` ([second paragraph](https://github.com/czyzby/gdx-lml/wiki/Your-first-LML-application#global-actions)) and an LML comment for the `{element}`.

**Comment by jpsgamboa** [comment]
Yes I believe you covered all of my points. Thanks.

**Comment by czyzby** [comment]
Great. I think that LML is ready for the `1.6`, but I'm going to wait a bit (~week?) until `VisUI 1.0.2` is released. A new LibGDX version would be nice, but that's a big "maybe".

I appreciate your input and using of snapshot versions - if you encounter any bugs or missing functionalities, make sure to create an issue. May I ask if you use the new `LmlApplicationListener` in your project?

**Comment by jpsgamboa** [comment]
Thanks for everything. I'm not using it now, but I still might. I'm still testing all the possible ways I can use LML.
